collector/src/main/java/com/bitmovin/analytics/retryBackend/RetryBackend.kt

```diff
@@ -0,0 +1,162 @@
+package com.bitmovin.analytics.retryBackend
+
+import android.os.Handler
+import android.os.SystemClock
+import android.util.Log
+import com.bitmovin.analytics.data.AdEventData
+import com.bitmovin.analytics.data.Backend
+import com.bitmovin.analytics.data.EventData
+import com.bitmovin.analytics.utils.Util.MAX_RETRY_SAMPLES
+import com.bitmovin.analytics.utils.Util.MAX_RETRY_TIME
+import java.io.IOException
+import java.lang.Exception
+import java.net.ConnectException
+import java.net.SocketTimeoutException
+import java.net.UnknownHostException
+import java.util.Calendar
+import java.util.Date
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.math.pow
+import okhttp3.Call
+import okhttp3.Callback
+import okhttp3.Response
+import okhttp3.internal.http2.StreamResetException
+
+class RetryBackend(private val next: Backend, val handler: Handler) : Backend {
+
+    private val TAG = "RetryBackend"
+    private val lock = ReentrantLock()
+    private var retrySamplesSet = sortedSetOf<RetrySample>()
+    private var retryToken: Date? = null
+
+    override fun send(eventData: EventData, callback: Callback?) {
+        scheduleSample(RetrySample(eventData, null, 0, Date(), 0))
+    }
+
+    override fun sendAd(eventData: AdEventData, callback: Callback?) {
+        scheduleSample(RetrySample(null, eventData, 0, Date(), 0))
+    }
+
+    private fun scheduleSample(retrySample: RetrySample) {
+
+        Log.d(TAG, "sending sample retry ${retrySample.retry}")
+
+        if (retrySample.eventData != null) {
```

Comment:
@Tigraine not sure about that part. Maybe we should separate between `ads` and normal samples and not have it in one list?
@KulenovicInes maybe make the `RetrySample` generic with a `data` property, and then handle it accordingly. This would be easier to extend in the future, now it's really specific to `EventData` and `AdEventData`.

Reply:
Added generic parameter